scripts/fit_pgd.py

```diff
@@ -0,0 +1,165 @@
+  loss = tf.reduce_sum((weight * (1 - pk / target_pk / rescale_factor))**2)
+  if return_pk:
+    return loss, pk
+  else:
+    return loss
+
+
+def main(_):
+  cosmology = flowpm.cosmology.Planck15()
+  # Create a simple Planck15 cosmology without neutrinos, and makes sure sigma8
+  # is matched
+  nbdykit_cosmo = Cosmology.from_astropy(Planck15.clone(m_nu=0 * u.eV))
+  nbdykit_cosmo = nbdykit_cosmo.match(sigma8=cosmology.sigma8.numpy())
+
+  # Compute the k vectora that will be needed in the PGD fit
+  k, _ = flowpm.power_spectrum(
+      tf.zeros([1] + [FLAGS.nc] * 3),
+      boxsize=np.array([FLAGS.box_size] * 3),
+      kmin=np.pi / FLAGS.box_size,
+      dk=2 * np.pi / FLAGS.box_size)
+
+  # Create some initial conditions
+  klin = tf.constant(np.logspace(-4, 1, 512), dtype=tf.float32)
+  pk = linear_matter_power(cosmology, klin)
+  pk_fun = lambda x: tf.cast(
+      tf.reshape(
+          interpolate.interp_tf(
+              tf.reshape(tf.cast(x, tf.float32), [-1]), klin, pk), x.shape), tf.
+      complex64)
+
+  initial_conditions = flowpm.linear_field(
+      [FLAGS.nc, FLAGS.nc, FLAGS.nc],
+      [FLAGS.box_size, FLAGS.box_size, FLAGS.box_size],
+      pk_fun,
+      batch_size=FLAGS.batch_size)
+
+  initial_state = flowpm.lpt_init(cosmology, initial_conditions, FLAGS.a_init)
+  stages = np.linspace(FLAGS.a_init, 1., FLAGS.nsteps, endpoint=True)
+
+  print('Starting simulation')
+  # Run the Nbody
+  states = flowpm.nbody(
+      cosmology,
+      initial_state,
+      stages, [FLAGS.nc, FLAGS.nc, FLAGS.nc],
+      pm_nc_factor=FLAGS.B,
+      return_intermediate_states=True)
+  print('Simulation done')
+
+  # Initialize PGD params
+  alpha = tf.Variable([FLAGS.alpha0], dtype=tf.float32)
+  scales = tf.Variable([FLAGS.kl0, FLAGS.ks0], dtype=tf.float32)
+  pgdparams = tf.concat([alpha, scales], 0)
+  optimizer = tf.keras.optimizers.Adam(learning_rate=FLAGS.learning_rate)
+
+  params = []
+  scale_factors = []
+  # We begin by fitting the last time step
+  for j, (a, state) in enumerate(states[::-1]):
+    # Let's compute the target power spectrum at that scale factor
+    target_pk = HalofitPower(nbdykit_cosmo, 1. / a - 1.)(k)
+
+    for i in range(FLAGS.niter if j == 0 else FLAGS.niter_refine):
+      optimizer.minimize(
+          partial(pgd_loss, alpha, scales, state, target_pk), [alpha] if
+          (FLAGS.fix_scales and j > 0) else [alpha, scales])
+
+      if i % 10 == 0:
+        loss, pk = pgd_loss(alpha, scales, state, target_pk, return_pk=True)
+        if i == 0:
+          pk0 = pk
+        print("step %d, loss:" % i, loss)
+    params.append(np.concatenate([alpha.numpy(), scales.numpy()]))
+    scale_factors.append(a)
+    print("Fitted params (alpha, kl, ks)", params[-1])
+
+    plt.loglog(k, target_pk, "k")
+    plt.loglog(k, pk0, ':', label='starting')
+    plt.loglog(k, pk, '--', label='after n steps')
+    plt.grid(which='both')
+    plt.savefig('PGD_fit_%0.2f.png' % a)
+    plt.close()
+
+  pickle.dump(
```

Comment:
Should we not be saving the boxsize and the grid size here as well?
Maybe also \omega_m and \sigma_8, but that depends on whether if we want to change cosmology/move to parametric form of the parameters at some point in the future.

Reply:
yeah that's a good point, it really doesn't hurt to have more meta info

Reply:
we are adding cosmological parameters as well as size and resolution of the simulation